packages/core/src/html/UlNodeProcessor.ts

```diff
@@ -0,0 +1,265 @@
+    'flex-shrink': '0',
+    height: '100%',
+  });
+}
+
+function updateNodeStyle(node: NodeOrText) {
+  const nodeCheerio = cheerio(node);
+
+  nodeCheerio.css({
+    'list-style-type': 'none',
+    'padding-inline-start': '0px',
+  });
+}
+
+const getIconAttributes = (node: MbNode, defaultIcon?: IconAttributes): IconAttributes | null => {
+  // If node.attribs doesn't exist, return null immediately
+  if (!node.attribs || (defaultIcon?.icon === undefined && node.attribs.icon === undefined)) {
+    return null;
+  }
+
+  // Prepare an object with default properties
+  const iconAttributes: IconAttributes = {
+    icon: defaultIcon?.icon,
+    width: defaultIcon?.width,
+    height: defaultIcon?.height,
+    size: defaultIcon?.size,
+    className: defaultIcon?.className,
+  };
+
+  // Overwrite defaults with the actual attributes from node if they exist
+  return {
+    ...iconAttributes,
+    icon: node.attribs.icon !== undefined ? node.attribs.icon : iconAttributes.icon,
+    width: node.attribs['i-width'] !== undefined ? node.attribs['i-width'] : iconAttributes.width,
+    height: node.attribs['i-height'] !== undefined ? node.attribs['i-height'] : iconAttributes.height,
+    size: node.attribs['i-size'] !== undefined ? node.attribs['i-size'] : iconAttributes.size,
+    className: node.attribs['i-class'] !== undefined
+      ? node.attribs['i-class']
+      : iconAttributes.className,
+  };
+};
+
+const updateUlNode = (node: MbNode, icon: IconAttributes) => {
+  if (get(node, 'attribs') && icon) {
+    node.attribs.icon = icon.icon ?? node.attribs.icon;
+    node.attribs['i-width'] = icon.width ?? node.attribs['i-width'];
+    node.attribs['i-height'] = icon.height ?? node.attribs['i-height'];
+    node.attribs['i-size'] = icon.size ?? node.attribs['i-size'];
+    node.attribs['i-class'] = icon.className ?? node.attribs['i-class'];
+  }
+};
+
+const deleteAttributes = (node: MbNode, attributes: string[]) => {
+  attributes.forEach((attr) => {
+    delete node.attribs[attr];
+  });
+};
+
+function updateLi(child: NodeOrText, iconAttributes: IconAttributes) {
+  const childNode = child as MbNode;
+
+  const curLiIcon = getIconAttributes(childNode, iconAttributes);
+
+  deleteAttributes(childNode, ICON_ATTRIBUTES);
+
+  // Create a new div and span
+  const div = cheerio('<div></div>');
+  const iconSpan = createIconSpan(curLiIcon!);
+
+  div.append(cheerio(child.children).remove());
+
+  // Append iconSpan and div to the child
+  cheerio(child).append(iconSpan);
+  cheerio(child).append(div);
+
+  cheerio(child).css({
+    display: 'flex',
+  });
+}
+
+function isRelevantNode(node: NodeOrText): boolean {
+  return 'name' in node && node.name === 'ul' && Boolean(node.children);
+}
+
+export function waterfallModel(node: NodeOrText) {
+  // Check if the node has `isIconListProcessed`. If yes, delete it and return.
+  const nodeAsMbNode = node as MbNode;
+  if (nodeAsMbNode.attribs.isIconListProcessed === 'true') {
```

Comment:
https://github.com/MarkBind/markbind/pull/2316#discussion_r1264700873

Doing the check here instead of `processUlNode` splits the processing across the original traversal process in `NodeProcessor -> second half of processUlNode` and here in `waterfallModel` which is difficult to follow.

Could we unify the processing flows? Something like

```js
// pseudo code

// UlNodeProcessor
function processUlNode(ulNode, currentLevelWaterfallState = {})
  if ulNode is processed
    stop

  mark ulNode processed

  nextLevelWaterfallState = { ... } // mutate this when processUlNode is called recursively. never change the reference

  for all li
    ... process with currentLevelWaterfallState ...
    ... update currentLevelWaterfallState ...

    if li has ul child
      call processUlNode(ul child, nextLevelWaterfallState) recursively


// NodeProcessor - no change
case 'ul':
  processUlNode(ulNode)
```

Reply:
Corrected to handle disjointed nested lists:

```js
// pseudo code

// UlNodeProcessor
function processUlNode(ulNode, level = 1, levelWaterfallStates = {})
  if ulNode is processed
    stop

  mark ulNode processed

  // if it does not exist create it
  // similar idea to iconAttrs in the dfs function currently
  currentLevelWaterfallState = levelWaterfallStates[level]

  for all li
    ... process with currentLevelWaterfallState ...
    ... update currentLevelWaterfallState ...

    if li has ul child
      call processUlNode(ul child, level + 1, levelWaterfallStates) recursively


// NodeProcessor - no change
case 'ul':
  processUlNode(ulNode)
```

Reply:
In short, pull the current implementation from

```
// processUlNode
  let defaultIcon = getIconAttributes(ulNode);
  ...
  return ulNode;
```

into `waterfallModel/dfs`